doc/website/blog/2022-03-14-ent-hooks-examples.md

```diff
@@ -0,0 +1,434 @@
+
+We start with a simple test
+```go
+func TestDogNameValidationHook(t *testing.T) {
+	ctx := context.Background()
+	c := enttest.Open(t, dialect.SQLite,
+		"file:TestSchemaConfHooks?mode=memory&cache=shared&_fk=1",
+	)
+	u := c.User.Create().SetName("Yoni").
+		SetConnectionString("mysql://root:pass@localhost:3306)").
+		SaveX(ctx)
+	_, err := c.Dog.Create().SetName("Yolo").SetOwner(u).Save(ctx)
+	require.Error(t, err)
+}
+```
+
+In this case, you can see that since "Yoni" and "Yolo" share the same 2 letters "Yo" the action fails.
+
+#### Recipe:
+
+First, register the hook:
+```go
+// Hooks of the Dog.
+func (Dog) Hooks() []ent.Hook {
+	return []ent.Hook{		
+		hook.If(validateName,
+			hook.HasFields(dog.FieldOwnerID),
+		),
+	}
+}
+```
+
+Since we want our hook to call every time we change the owner (to make sure the name is matched against), I am using an
+[edge field](https://entgo.io/docs/schema-edges#edge-field), this provides the hooks a trigger on when the edge's id changes.
+
+The hook:
+```go
+func validateName(next ent.Mutator) ent.Mutator {
+	return hook.DogFunc(func(ctx context.Context, m *gen.DogMutation) (ent.Value, error) {
+		owID, ok := m.OwnerID()
+		if !ok {
+			return next.Mutate(ctx, m)
+		}
+		owner, err := m.Client().User.Query().Where(user.ID(owID)).Only(ctx)
+		if err != nil {
+			return next.Mutate(ctx, m)
+		}
+		dn, ok := m.Name()
+		if !ok {
+			return next.Mutate(ctx, m)
+		}
+		if owner.Name[0:1] == dn[0:1] {
+			return nil, errors.New("invalid dog name")
+		}
+		return next.Mutate(ctx, m)
+	})
+}
+```
+A simple name validator and if we see a problem we cancel the mutation by returning an error before ```next.Mutate(ctx,m)```.
+
+### Offloading Long Computations:
+
+Some cases, we have fields in our database that require long computation, we would like to make sure that this data is always updated once
+a change in the data requires it.
+
+A test for it will look like this:
+```go
+func TestCacheHook(t *testing.T) {
+	ctx := context.Background()
+	cs := cache.NewSyncer()
+	c := enttest.Open(t, dialect.SQLite,
+		"file:TestSchemaConfHooks?mode=memory&cache=shared&_fk=1",
+		enttest.WithOptions(ent.CacheSyncer(cs)),
+	)
+	cs.Start(ctx, c)
+	cl := c.Cache.Create().SetWalks(-1).SaveX(ctx)
+	d := c.Dog.Create().SetName("Karashindo").SaveX(ctx)
+	u := c.User.Create().SetName("Yoni").
+		SetCache(cl).
+		AddPets(d).
+		SetConnectionString("mysql://root:pass@localhost:3306)").
+		SaveX(ctx)
+	c.Dog.UpdateOne(d).SetName("Fortuna").ExecX(ctx)
+	cs.Close()
+	cl = u.QueryCache().OnlyX(ctx)
+	require.True(t, cl.Walks > 0)
+}
+```
+This test is a bit more complicated so lets review it line by line:
+First, I create a new ```cache syncer``` (remember we have a cache entity in our schema).
+
+This ```Cache Syncer``` can do heavy computations  while running in a different  context.
+We create an ent client and inject our ```Cache Syncer```, it is disabled on default.
```

Comment:
Explain this and link to the docs. Not a widely known feature I think. The dependency injection I mean.

Reply:
👍